tuna/miopen/celery_tuning/tuning.py

```diff
@@ -0,0 +1,390 @@
+    except kombu.exceptions.OperationalError as k_err:
+      LOGGER.error('Redis error ocurred: %s', k_err)
+      return False
+
+  global DBT  #pylint: disable=global-variable-undefined
+  DBT = MIOpenDBTables(session_id=library.args.session_id,
+                       config_type=library.args.config_type)
+
+  fdb_attr = [column.name for column in inspect(DBT.find_db_table).c]
+  fdb_attr.remove("insert_ts")
+  fdb_attr.remove("update_ts")
+
+  f_vals = library.get_f_vals(Machine(local_machine=True),
+                              range(0),
+                              tuning=True)
+  kwargs = library.get_kwargs(0, f_vals, tuning=True)
+
+  return worker_type, kwargs, fdb_attr, q_name
+
+
+#pylint: disable=too-many-locals
+def tune(library, job_batch_size=1000):
+  """tuning loop to spin out celery tasks"""
+
+  try:
+    worker_type, kwargs, fdb_attr, q_name = prep_tuning(library)
+  except ValueError as verr:
+    LOGGER.error(verr)
+    return False
+
+  #if enqueue_only is False, we only launch the workers
+  if not library.args.enqueue_only:
+    return True
+
+  res_set = ResultSet([])
+  start = time.time()
+
+  with DbSession() as session:
+    while True:
+      try:
+        job_list = []
+        #get all the jobs from mySQL
+        job_list = library.get_jobs(session, library.fetch_state,
+                                    library.set_state, library.args.session_id,
+                                    job_batch_size)
+
+        for i in range(0, len(job_list), job_batch_size):
+          batch_jobs = job_list[i:min(i + job_batch_size, len(job_list))]
+          if library.args.fin_steps:
+            entries = library.compose_work_objs_fin(session, batch_jobs,
+                                                    library.dbt)
+
+          entries = library.compose_work_objs_fin(session, job_list,
+                                                  library.dbt)
+          serialized_jobs = serialize_chunk(entries)
+
+          #build context for each celery task
+          for job, config in serialized_jobs:
+            context = {
+                'job': job,
+                'config': config,
+                'worker_type': worker_type,
+                'arch': library.dbt.session.arch,
+                'num_cu': library.dbt.session.num_cu,
+                'kwargs': kwargs,
+                'fdb_attr': fdb_attr
+            }
+
+            #calling celery task, enqueuing to celery queue
+            res_set.add(celery_enqueue.apply_async((context,), queue=q_name))
+
+        if not job_list:
+          if not res_set:
+            return False
+          LOGGER.info('All tasks added to queue')
+          break
+      except KeyboardInterrupt:
+        LOGGER.error('Keyboard interrupt caught, draining results queue')
+        session.rollback()
+        stop_active_workers()
+        results_gather(res_set, worker_type)
+        purge_queue([q_name], LOGGER)
+
+  results_gather(res_set, worker_type)
```

Comment:
Will be preferable to have the async result gathering started before Tuna begins to fill the celery queue. So that Tuna isn't waiting on all work to be added before it starts processing results.

Reply:
correct, i have changes to push where this is moved before the get_jobs gets called.